Bottom: d5b58b706c00e9a331b77b231f39fedba4b13997
Top:    d48c04ee53e0f2a4beeccd166e8a4d9abaa61c48
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-10 15:30:37 +0000

Refresh of manually-count-the-vertices-in

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 374fe5e..9c5aa92 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -649,8 +649,22 @@ hidgl_fill_pcb_polygon (PolygonType *poly)
   /* Walk the polygon structure, counting vertices */
   piece = poly->Clipped;
 //  do {
-    for (contour = piece->contours; contour != NULL; contour = contour->next)
+    for (contour = piece->contours; contour != NULL; contour = contour->next) {
+      int contour_count_manual = 0;
+      vnode = &contour->head;
+      do {
+        contour_count_manual++;
+      } while ((vnode = vnode->next) != &contour->head);
+
+      if (contour_count_manual != contour->Count) {
+        fprintf (stderr, "CRAPPY POLYGON ROUTINES MISCOUNTED NODES, FIXING\n");
+        fprintf (stderr, "Manual count was %i, contour claimed %i\n",
+                 contour_count_manual, contour->Count);
+        contour->Count = contour_count_manual;
+      }
+
       vertex_count += contour->Count;
+    }
 //  } while ((piece = piece->f) != poly->Clipped);
 
   vertices = malloc (sizeof(GLdouble) * vertex_count * 3);
